### PR TITLE
Emotes on alertpage

### PR DESCRIPTION
**Pull request by radipiz**
Hi, first time contributing. I've cleaned my code and created a hopefully comprehensive commit history.
You probably know the flying emotes from other proprietary overlays.

I've implemented the following features:
* [Alerts] Support to display emotes (local, Emoji, BTTV, FFZ, Twitch Native + animated v2 Emotes)
* [Alerts] Support to play multiple media at once or sequential
* [Alerts] Support to run macros triggering emotes, sounds or videos (currently there is no frontend or action implemented to define or trigger these; I'm using a ui which is about 80% finished but probably too much for this PR)

I've added the following libraries:
* [com.vdurmont.emoji-java](https://github.com/vdurmont/emoji-java) (MIT License)
* [Keyframes](https://github.com/Keyframes/Keyframes) (MIT License)
* [animate.style](https://animate.style/) (MIT License)

Regarding my changes:
* `javascript-source/handlers/emojiHandler.js`: Handler module to process emoji and triggers an event on alertspollsocket
* `javascript-source/handlers/emotesBttvHandler.js`: Handler module to process emotes from BTTV and triggers an event on alertspollsocket
* `javascript-source/handlers/emotesFfzHandler.js`: Handler module to process emotes from FFZ and triggers an event on alertspollsocket
* `javascript-source/handlers/emotesTwitchHandler.js`: Handler module to trigger an alert on alertspollsocket for emotes from Twitch 
* ` javascript-source/handlers/emotesHandler.js`: Added FFZ and BTTV emotes to build a local emote cache for fast detection in chat messages
* `resources/config/clips/quak.webm`: Some demo file that can be shown on alerts page (source internet, widely spread video clip)
* `resources/config/clips/quak.webm`: Some demo file that can be shown on alerts page (source: i drew it)
* `resources/web/alerts/css/animate.min.css`: animate.style framework for animations with css used for the flying emotes
* `resources/web/alerts/index.html`:  Added elements for playing videos and emotes (and reformatted the file)
* `resources/web/alerts/js/index.js`: 
   * Pulled some constants into variables to make them easier to use
   * Rewrote queue handling so it can process a greater variety of events and play them simultaneously
   * Added flying emotes feature
   * Added handling for video clips
   * Added a macro feature to let different events play as a little macro
 * `resources/web/alerts/js/keyframes.dist.js`: Keyframes resource file used to generate the animations with js and css
 * `resources/web/panel/index.html`: Added a configuration page on the panel to configure the emote handlers
     * emotesAndEmoji page allows to enable and disable certain providers
 * `source/com/gmt2001/PathValidator.java`: Added clips and emotes path to allowed paths so locally stored content can be served
 * `source/com/gmt2001/httpwsserver/HTTPWSServer.java`: Added clips and emotes path to allowed paths so locally stored content can be served
 * `source/com/gmt2001/httpwsserver/HttpSslRedirectHandler.java`: same
 * `source/tv/phantombot/httpserver/HTTPNoAuthHandler.java`: same
 * `source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java`: same
 * `source/tv/phantombot/panel/WsAlertsPollsHandler.java`
    * Added methods to send emotes event in different overloads
    * Added method to play a video clip
    * added a method to stop all playing
    * added a method to play a list of commands (macro)
    * changed `alertImage` and `triggerAudioPanel` to contain the `ignoreIsPlaying` flag (Someone on the Discord asked for such a feature and I casually implemented it)
 * `source/tv/phantombot/scripts/handler/EmojiProcessor.java`: Helper class to process emojis to their codepoints

I hope it's understandable and welcome what I did. Once this is accepted, I have another feature that is built on this base.
If you have questions or suggestions, just contact me. I'm also on the Discord server for short notices.

### CODE REVIEW

File: resources/web/alerts/js/index.js [hunk 1]
```diff
@@ -282,6 +355,190 @@ $(function () {
...
+            return;
+        }
+
+        // scaling of the emote (by width)
+        const size = 112;
+
+        const browserSafeId = emoteId.replace(/\W/g, '');
+
+        // a pseudo unique id to make sure, the keyframe names won't interfere each other
+        const uniqueId = `${Date.now()}${Math.random().toString(16).substr(2, 8)}`
```
Comment (gmt2001) [inline resources/web/alerts/js/index.js]: Codacy Production / Codacy Static Code Analysis

resources/web/alerts/js/index.js#L388

Missing semicolon.

File: resources/web/alerts/js/index.js [hunk 2]
```diff
@@ -282,6 +355,190 @@ $(function () {
...
+            return;
+        }
+        let defaultPath = '/config/clips';
+        let filename = json.filename;
+        let duration = json.duration || -1;
+        let fullscreen = json.fullscreen || false;
+        let volume = getOptionSetting(CONF_VIDEO_CLIP_VOLUME, '0.8');
+
+        let video = document.createElement('video');
+        video.src = `${defaultPath}/${filename}`
```
Comment (gmt2001) [inline resources/web/alerts/js/index.js]: Codacy Production / Codacy Static Code Analysis

resources/web/alerts/js/index.js#L514

Missing semicolon.

File: source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     * },
+     * local: {
+     * ...
+     * },
+     * shared: {
+     * ...
+     * }
+     * }
+     */
+    private NativeObject bttvEmotes;
```
Comment (gmt2001) [inline source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java]: Codacy Production / Codacy Static Code Analysis

source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java#L28

Private field 'bttvEmotes' could be made final; it is only initialized in the declaration or constructor.

File: source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+     * },
+     * local: {
+     * ...
+     * },
+     * shared: {
+     * ...
+     * }
+     * }
+     */
+    private NativeObject ffzEmotes;
```
Comment (gmt2001) [inline source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java]: Codacy Production / Codacy Static Code Analysis

source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java#L45

Private field 'ffzEmotes' could be made final; it is only initialized in the declaration or constructor.

File: source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    public EmotesCacheUpdatedEvent(NativeObject bttvEmotes, NativeObject ffzEmotes) {
```
Comment (gmt2001) [inline source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java:47]: Codacy Production / Codacy Static Code Analysis

source/tv/phantombot/event/emotes/EmotesCacheUpdatedEvent.java#L47

It is a good practice to call super() in a constructor

File: resources/web/panel/pages/keywords/emotesAndEmoji.html [hunk 6]
```diff
@@ -0,0 +1,106 @@
+<!--
+    Copyright (C) 2016-2021 phantombot.github.io/PhantomBot
```
Comment (gmt2001) [inline resources/web/panel/pages/keywords/emotesAndEmoji.html]: Update Copyright to 2022

File: resources/web/panel/js/pages/keywords/emotesAndEmoji.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2016-2021 phantombot.github.io/PhantomBot
```
Comment (gmt2001) [inline resources/web/panel/js/pages/keywords/emotesAndEmoji.js]: Update copyright to 2022

File: resources/web/panel/js/pages/keywords/emotesAndEmoji.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+page = {
```
Comment (gmt2001) [inline resources/web/panel/js/pages/keywords/emotesAndEmoji.js:18]: Global variable declared, enclose local scoped code inside lambda function and explicitly export only as necessary
```
$(function () {
    //Code here
});
```

File: javascript-source/handlers/emojiHandler.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2016-2021 phantombot.github.io/PhantomBot
```
Comment (gmt2001) [inline javascript-source/handlers/emojiHandler.js]: Update copyright to 2022

File: javascript-source/handlers/emojiHandler.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        let emojiCounts = Packages.tv.phantombot.scripts.handler.EmojiProcessor.groupByCount(codepoints);
+
+        if (emojiCounts.size() > 0) {
+            emojiCounts.forEach((codepoint, count) =>
+                $.alertspollssocket.triggerEmote(codepoint, count, emoteProvider)
+            );
+        }
+    });
+
+    $.bind('initReady', function () {
```
Comment (gmt2001) [inline javascript-source/handlers/emojiHandler.js]: Remove unused hooks

**Review by gmt2001 — CHANGES_REQUESTED** [review]
There were a few minor things that were triggered by lint

Also, can you confirm that this was tested on the latest bot _without_ using the `rhino_es6` option?

**Comment by radipiz** [comment]
Good find with the Rhino ES5 engine! I was using Rhino ES6 before it was a configurable option. There are some hints that my code is older because I implemented it for myself back then and used it with my configuration for more than a year now.

I hope, I've now solved all the issues.

**Review by gmt2001 — APPROVED** [review]

**Pull request merged**